cmd/util/cmd/find-trie-root/cmd.go

```diff
@@ -0,0 +1,379 @@
+
+	defer writer.Close()
+
+	sr, err := prometheusWAL.NewSegmentsRangeReader(log.Logger, prometheusWAL.SegmentRange{
+		Dir:   dir,
+		First: segment,
+		Last:  segment,
+	})
+	if err != nil {
+		return "", fmt.Errorf("cannot create WAL segments reader: %w", err)
+	}
+
+	defer sr.Close()
+
+	reader := prometheusWAL.NewReader(sr)
+
+	for reader.Next() {
+		record := reader.Record()
+		operation, _, update, err := wal.Decode(record)
+		if err != nil {
+			return "", fmt.Errorf("cannot decode LedgerWAL record: %w", err)
+		}
+
+		switch operation {
+		case wal.WALUpdate:
+
+			bytes := wal.EncodeUpdate(update)
+			_, err = writer.Log(bytes)
+			if err != nil {
+				return "", fmt.Errorf("cannot write LedgerWAL record: %w", err)
+			}
+
+			rootHash := update.RootHash
+
+			if rootHash.Equals(expectedRoot) {
+				log.Info().Msgf("found expected trie root hash %v, finish writing", rootHash)
+				return newSegmentFile, nil
+			}
+		default:
+		}
+
+		err = reader.Err()
+		if err != nil {
+			return "", fmt.Errorf("cannot read LedgerWAL: %w", err)
+		}
+	}
+
+	return "", fmt.Errorf("finish reading all segment files from %d to %d, but not found", segment, segment)
+}
+
+func checkFolderNotExistOrEmpty(folderPath string) (bool, error) {
+	// Check if the folder exists
+	info, err := os.Stat(folderPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	// Check if the path is a directory
+	if !info.IsDir() {
+		return false, fmt.Errorf("The path is not a directory.")
+	}
+
+	// Check if the folder is empty
+	files, err := os.ReadDir(folderPath)
+	if err != nil {
+		return false, fmt.Errorf("Cannot read the folder.")
+	}
+
+	return len(files) == 0, nil
+}
+
+// backup new wals before replacing
+func backupRollbackedWALsAndMoveLastSegmentFile(
+	segment int, walDir, backupDir string, newSegmentFile string) error {
+	// making sure the backup dir is empty
+	empty, err := checkFolderNotExistOrEmpty(backupDir)
+	if err != nil {
+		return fmt.Errorf("cannot check backup directory: %w", err)
+	}
+
+	if !empty {
+		return fmt.Errorf("backup directory %s is not empty", backupDir)
+	}
+
+	// Create the backup directory
+	err = os.MkdirAll(backupDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("cannot create backup directory: %w", err)
+	}
+
+	first, last, err := prometheusWAL.Segments(walDir)
+	if err != nil {
+		return fmt.Errorf("cannot get segments: %w", err)
+	}
+
+	if segment < first {
+		return fmt.Errorf("segment %d is less than the first segment %d", segment, first)
+	}
+
+	// backup all the segment files that have higher number than the given segment, including
+	// the segment file itself, since it will be replaced.
+	for i := segment; i <= last; i++ {
+		segmentFile := prometheusWAL.SegmentName(walDir, i)
+		backupFile := prometheusWAL.SegmentName(backupDir, i)
+
+		log.Info().Msgf("backup segment file %s to %s, %v/%v", segmentFile, backupFile, i, last)
+		err := os.Rename(segmentFile, backupFile)
```

Comment:
Do we want to use `copyFile()` here in case backup directory is in a different device?

Reply:
Yeah, I thought about this. If I use copy file, it will take some time, I think we most likely would move within the same device. If a different device is used, then it fails, they can retry again with a different path on the same device. 

I could also check if they are different device, and then use copy and remove the old files, but that would be too much code. I decided to keep it simple.